### PR TITLE
Value type null-restricted array support

### PR DESCRIPTION
**Pull request by theresa-m**
This change is being resubmitted from https://github.com/eclipse-openj9/openj9/pull/19911 with the only change being the most recent commit. The commit prevents infinite looping in GC_ClassLoaderClassesIterator::nextArrayClass by ensuring `_iterateArrayClass ` is reset to NULL at the end of a loop.
Personal builds:
http://vmfarm.rtp.raleigh.ibm.com/build_info.php?build_id=77569 
https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/23806/
- Discussed the extended failure SyntheticGCWorkload_concurrentSlackAuto_10M_J9_0 with Dmitri, it is an unreliable test that is not related to this change.

---
Please note this pull request adds support to create a null-restricted array but not a flattened array. That will come in a later change.

- create a null-restricted array field for every RAM class if one doesn't exist when a null-restricted array is created (see https://github.com/eclipse-openj9/openj9/issues/19460#issue-2284036228)
- implement JVM_IsNullRestrictedArray and JVM_NewNullRestrictedArray
- Update exception thrown when null is set to a null-restricted array from NullPointerException to ArrayStoreException as described in https://openjdk.org/jeps/8316779

Related: https://github.com/eclipse-openj9/openj9/issues/17340
Fixes: https://github.com/eclipse-openj9/openj9/issues/19460

After these changes:

- ValueTypeArrayTests have been enabled and pass https://github.com/eclipse-openj9/openj9/issues/19460#issuecomment-2163847738
- ValueTypeTestsJIT still has some failures https://github.com/eclipse-openj9/openj9/issues/19708
- ValueTypeArrayTestsJIT still has some failures https://github.com/eclipse-openj9/openj9/issues/19913
- ValueTypeSystemArraycopyTests pass for Xint variations https://github.com/eclipse-openj9/openj9/issues/19914

### CODE REVIEW

**Review by dmitripivkine — APPROVED** [review]

File: runtime/gc_structs/ClassLoaderClassesIterator.cpp [hunk 1]
```diff
@@ -99,6 +102,60 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
...
+			} else {
+				_iterateArrayClazz = NULL;
+				_arrayState = STATE_DONE;
+			}
+			break;
+		case STATE_ARRAYLIST:
+			if (NULL != _iterateArrayClazz) {
+				_iterateArrayClazz = _iterateArrayClazz->arrayClass;
+			} else {
+				_iterateArrayClazz = NULL;
```
Comment (keithc-ca) [inline runtime/gc_structs/ClassLoaderClassesIterator.cpp:144]: Please remove this redundant assignment (notice line 141).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/gc_structs/ClassLoaderClassesIterator.cpp (hunk 1 above)
Comment (theresa-m) [reply runtime/gc_structs/ClassLoaderClassesIterator.cpp]: I'm running another Java 8 build with this update for good measure http://vmfarm.rtp.raleigh.ibm.com/build_info.php?build_id=77729

**Review by keithc-ca — APPROVED** [review]

**Comment by hangshao0** [comment]
Jenkins test sanity,extended zlinuxval jdknext

**Comment by hangshao0** [comment]
Jenkins test sanity amac jdk21

**Review by hangshao0 — APPROVED** [review]

**Pull request merged**